### PR TITLE
[Feature] Add real-time node prediction inferrer

### PR DESCRIPTION
**Pull request by zhjwy9343**
*Issue #, if available:*

*Description of changes:*
This PR add a new real-time node prediction inferrer, `GSGnnNodePredictionRealtimeInferrer`. And modify the `GSgnnData` class for extracting edge features.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: python/graphstorm/dataloading/dataset.py [hunk 1]
```diff
@@ -425,6 +464,10 @@ def get_blocks_edge_feats(self, input_blocks, efeat_fields, device='cpu'):
         .. versionadded:: 0.4.0
             Add ``get_blocks_edge_feat`` in 0.4.0 to support edge features in message passing.
 
+        .. versionchanged:: 0.5
+            Moved this function out of `GSgnnData` class in v0.5 to support real-time node-level
```
Comment (classicsong) [inline python/graphstorm/dataloading/dataset.py:468]:
```suggestion
            Moved the implementatino of this function out of `GSgnnData` class in v0.5 to support real-time node-level
```

File: python/graphstorm/dataloading/dataset.py [hunk 2]
```diff
@@ -425,6 +464,10 @@ def get_blocks_edge_feats(self, input_blocks, efeat_fields, device='cpu'):
...
+            inference on Amazon SageMaker endpoints.
```
Comment (classicsong) [inline python/graphstorm/dataloading/dataset.py:469]:
```suggestion
            inference.
```

File: python/graphstorm/inference/__init__.py [hunk 3]
```diff
@@ -17,7 +17,7 @@
 """
 
 from .lp_infer import GSgnnLinkPredictionInferrer
-from .np_infer import GSgnnNodePredictionInferrer
+from .np_infer import GSgnnNodePredictionInferrer, GSGnnNodePredictionRealtimeInferrer
```
Comment (classicsong) [inline python/graphstorm/inference/__init__.py:20]:
```suggestion
from .np_infer import (GSgnnNodePredictionInferrer, 
                                       GSGnnNodePredictionRealtimeInferrer)
```

File: python/graphstorm/inference/np_infer.py [hunk 4]
```diff
@@ -168,3 +171,90 @@ def infer(self, loader, save_embed_path, save_prediction_path=None,
             save_node_prediction_results(shuffled_preds, save_prediction_path)
         barrier()
         sys_tracker.check('save predictions')
+
+
+class GSGnnNodePredictionRealtimeInferrer(GSInferrer):
+    """ Inferrer for real-time node prediction tasks on SageMaker endpoints.
+
+    .. versionadded:: 0.5
+        Add `GSGnnNodePredictionRealtimeInferrer` class in v0.5 to support real-time node-level
```
Comment (classicsong) [inline python/graphstorm/inference/np_infer.py:180]:
```suggestion
        Add `GSGnnNodePredictionRealtimeInferrer` class to support real-time node-level
```

File: python/graphstorm/inference/np_infer.py [hunk 5]
```diff
@@ -168,3 +171,90 @@ def infer(self, loader, save_embed_path, save_prediction_path=None,
...
+        inference on Amazon SageMaker endpoints.
+
+    The real-time inferrer has three major differences from the node prediction offline inferrer:
+
+    1. directly setting and using a DGL graph as input to extract features.
```
Comment (classicsong) [inline python/graphstorm/inference/np_infer.py:185]: What do you mean by directly setting?

File: python/graphstorm/inference/np_infer.py [hunk 6]
```diff
@@ -168,3 +171,90 @@ def infer(self, loader, save_embed_path, save_prediction_path=None,
...
+    2. using DGL dataloader, instead of GS dataloaders that rely on DGL distributed graphs.
```
Comment (classicsong) [inline python/graphstorm/inference/np_infer.py:186]: Shall we wrap DGL dataloader with GSInferDataloaders? So we can define the APIs for custom RealtimeInferrer.

File: python/graphstorm/inference/np_infer.py [hunk 7]
```diff
@@ -168,3 +171,90 @@ def infer(self, loader, save_embed_path, save_prediction_path=None,
...
+
+        Parameters
+        ----------
+        g: DGLGraph
+            The inference graph data in the format of a DGL heterograph. For built-in inference
+            pipeline, this graph should be constructed by using methods in the
+            `gconstruct.construct_payload_graph.py` file.
+        dataloader: DGL dataloader
+            A DGL dataloader class for node prediction, e.g., `DataLoader`.
+        infer_ntypes: list of string or a string
```
Comment (classicsong) [inline python/graphstorm/inference/np_infer.py:212]: Can we make it a list of strings?

File: python/graphstorm/inference/np_infer.py [hunk 8]
```diff
@@ -168,3 +171,90 @@ def infer(self, loader, save_embed_path, save_prediction_path=None,
...
+        if isinstance(infer_ntypes, str):
+            infer_ntypes = [infer_ntypes]
+
+        # set model to be in the evaluation mode
+        self._model.eval()
+        # extract one mini-batch blocks using the given DGL dataloader
+        all_nodes = []
+        all_blocks = []
+        for input_nodes, _, blocks in dataloader:
+            all_nodes = input_nodes
```
Comment (classicsong) [inline python/graphstorm/inference/np_infer.py:242]: all_nodes.append?

You only store the last element from dataloader.

File: tests/unit-tests/test_inferrer.py [hunk 9]
```diff
@@ -1674,15 +1677,90 @@ def test_hgt_infer_lp4ef():
...
+
+    model = create_builtin_node_gnn_model(g, config, True)
+
+    # initialize a real-time np inferrer
+    inferrer = GSGnnNodePredictionRealtimeInferrer(model)
+
+    # initialize a DGL dataloader
+    sampler = MultiLayerFullNeighborSampler(config.num_layers)
+    dataloader = DataLoader(g, infer_nids, sampler,
+                            batch_size=10, shuffle=False,
```
Comment (classicsong) [inline tests/unit-tests/test_inferrer.py:1713]: Please also test the case when num_targets > 10

File: tests/unit-tests/test_inferrer.py [hunk 10]
```diff
@@ -1674,15 +1677,90 @@ def test_hgt_infer_lp4ef():
...
+    # test_hgt_infer_ec4ef()
+    # test_hgt_infer_lp4ef()
 
-    test_rgcn_infer_nc4ef()
-    test_rgcn_infer_ec4ef()
-    test_rgcn_infer_lp4ef()
-    test_hgt_infer_nc4ef()
-    test_hgt_infer_ec4ef()
-    test_hgt_infer_lp4ef()
+    test_realtime_infer_np()
```
Comment (classicsong) [inline tests/unit-tests/test_inferrer.py:1766]: remove __main__

File: python/graphstorm/inference/np_infer.py (hunk 6 above)
Comment (zhjwy9343) [reply python/graphstorm/inference/np_infer.py:186]: Yes, create a new `GSgnnRealtimeInferNodeDataLoader` class. Also allow users to use DGL `DataLoader` as the two classes return the same outputs.

File: python/graphstorm/inference/np_infer.py (hunk 7 above)
Comment (zhjwy9343) [reply python/graphstorm/inference/np_infer.py:212]: The format of a string and a list of string is compatible with GS's `target_ntype` argument. I would suggest to keep this compatibility.

File: python/graphstorm/dataloading/dataset.py [hunk 11]
```diff
@@ -154,6 +154,45 @@ def prepare_batch_edge_input(g, input_edges,
                 feat[etype] = th.cat(feats, dim=1)
     return feat
 
+def prepare_blocks_edge_feats(g, input_blocks, efeat_fields, device='cpu'):
```
Comment (thvasilo) [inline python/graphstorm/dataloading/dataset.py:157]: Is this change related to this PR?

File: python/graphstorm/dataloading/dataset.py [hunk 12]
```diff
@@ -425,6 +464,10 @@ def get_blocks_edge_feats(self, input_blocks, efeat_fields, device='cpu'):
         .. versionadded:: 0.4.0
             Add ``get_blocks_edge_feat`` in 0.4.0 to support edge features in message passing.
 
+        .. versionchanged:: 0.5
+            Moved the implementatino of this function out of `GSgnnData` class in v0.5 to support
+            real-time node-level inference.
```
Comment (thvasilo) [inline python/graphstorm/dataloading/dataset.py:469]: So we moved the function in order to allow it to be called without a GSgnnData object?

File: python/graphstorm/inference/np_infer.py [hunk 13]
```diff
@@ -17,6 +17,7 @@
 """
 import time
 import logging
+import dgl
```
Comment (thvasilo) [inline python/graphstorm/inference/np_infer.py:20]: DGL is a 3rd part import should go in the second import group

File: python/graphstorm/inference/np_infer.py (hunk 8 above)
Comment (thvasilo) [reply python/graphstorm/inference/np_infer.py:247]: Yes if we know the iterator only has one element we need to make that clear in the code. Maybe add an assertion as well

File: python/graphstorm/dataloading/dataloading.py [hunk 14]
```diff
@@ -1784,6 +1784,99 @@ def __len__(self):
...
                    self.unlabeled_dataloader.expected_idxs)
 
+class GSgnnRealtimeInferNodeDataLoader(GSgnnNodeDataLoaderBase):
+    """ Mini-batch dataloader for real-time node infernce task
+
+    .. versionadded:: 0.5
+        Add `GSgnnRealtimeInferNodeDataLoader` class to support node-level dataloader used for
+        real-time inference.
+
+    This dataloader wraps DGL's DataLoader for real-time inference in GraphStorm. It will use DGL's
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply python/graphstorm/dataloading/dataloading.py:1794]: Yes. Revised the doc string.

File: python/graphstorm/dataloading/dataloading.py [hunk 15]
```diff
@@ -1784,6 +1784,99 @@ def __len__(self):
...
+                 target_idx,
+                 num_layers=1):
+        assert isinstance(dataset, dgl.DGLGraph), ('The \"dataset\" should be a DGLGraph ' \
+            f'instance, but got {dataset}.')
+        assert isinstance(target_idx, dict), ('The \"target_idx\" should be a dictionary, ' \
+            f'but got {target_idx}.')
+        for ntype in target_idx:
+            assert ntype in dataset.ntypes, (f'node type {ntype} does not exist in the graph.')
+
+        # set dummy label fields using "label" string
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply python/graphstorm/dataloading/dataloading.py:1822]: Because the super class `GSgnnNodeDataLoaderBase` asks for a valid label field, if not provide or set as `None`, the `verify_label_field()` in `GSgnnNodeDataLoaderBase` will fail.

File: python/graphstorm/dataloading/dataset.py (hunk 11 above)
Comment (zhjwy9343) [reply python/graphstorm/dataloading/dataset.py:157]: Yes, the new `GSGnnNodePredictionRealtimeInferrer` class' `infer()` function need it to get features.

File: python/graphstorm/dataloading/dataset.py (hunk 12 above)
Comment (zhjwy9343) [reply python/graphstorm/dataloading/dataset.py:469]: Yes, the `GSGnnNodePredictionRealtimeInferrer` does not use `GSgnnData`, but a DGLGraph. So need this function to be independent.

File: python/graphstorm/inference/np_infer.py (hunk 13 above)
Comment (zhjwy9343) [reply python/graphstorm/inference/np_infer.py:20]: changed.

File: python/graphstorm/dataloading/dataloading.py [hunk 16]
```diff
@@ -1784,6 +1784,99 @@ def __len__(self):
...
+        for _, idx in target_idx.items():
+            if len(idx) > batch_size:
+                batch_size = len(idx)
+
+        # use a full neighbor sampler because it is unclear if callers have done sampling when
+        # building the subgraph payload.
+        sampler = dgl.dataloading.MultiLayerFullNeighborSampler(num_layers)
+
+        device = get_device() \
+            if is_distributed() and get_backend() == "nccl" else th.device('cpu')
```
[the remark this answers is not in the document]
Comment (zhjwy9343) [reply python/graphstorm/dataloading/dataloading.py:1849]: Replaced with using native torch cuda function to use one GPU only in a GPU instance for real-time inference. Distributed intialization may not work well.

File: python/graphstorm/inference/np_infer.py (hunk 8 above)
Comment (zhjwy9343) [reply python/graphstorm/inference/np_infer.py:247]: Yes. revised to use `.next()`

**Review by thvasilo — APPROVED** [review]
LGTM

File: python/graphstorm/inference/np_infer.py [hunk 17]
```diff
@@ -241,21 +241,26 @@ def infer(self,
         # set model to be in the evaluation mode
         self._model.eval()
         # extract one mini-batch blocks using the given dataloader
-        all_nodes = {}
-        all_blocks = []
-        for input_nodes, _, blocks in dataloader:
-            all_nodes = input_nodes
-            all_blocks = blocks
+        assert len(dataloader) == 1, ('Real-time inference do mini-batch computing once, but ' \
```
Comment (classicsong) [inline python/graphstorm/inference/np_infer.py:244]:
```suggestion
        assert len(dataloader) == 1, ('Real-time inference do single batch computation, but ' \
```

File: python/graphstorm/dataloading/dataloading.py [hunk 18]
```diff
@@ -1785,51 +1788,45 @@ def __len__(self):
...
-        # set dummy label fields using "label" string
-        label_field = "label"
+            assert ntype in g.ntypes, (f'node type {ntype} does not exist in the graph.')
 
-        super().__init__(dataset,
+        super().__init__(g,
                          target_idx,
                          fanout=[-1],
-                         label_field=label_field,
+                         label_field="label",
```
Comment (classicsong) [inline python/graphstorm/dataloading/dataloading.py:1825]: why this is a constant value?

File: tests/unit-tests/test_dataloading.py [hunk 19]
```diff
@@ -2667,24 +2667,22 @@ def test_realtime_infer_node_dataloader():
...
-    for input_nodes, seeds, blocks in dataloader:
-        assert 'n0' in input_nodes
-        assert input_nodes['n0'].shape[0] == len(all_src_idx)
-        assert 'n1' in input_nodes
-        
-        assert input_nodes['n1'].shape[0] == batch_size
-        assert 'n1' in seeds
-        assert len(blocks) == 1
-        all_nodes.append(seeds['n1'])
+    # should only has one mini batch
```
Comment (classicsong) [inline tests/unit-tests/test_dataloading.py:2670]: You should also check the len of dataloader

**Review by classicsong — APPROVED** [review]

File: python/graphstorm/dataloading/dataloading.py (hunk 18 above)
Comment (zhjwy9343) [reply python/graphstorm/dataloading/dataloading.py:1825]: The base class `GSgnnNodeDataLoaderBase` asks for a valid l`abel_field`. So, here give it a dummy name. Added some comments to explain.

File: tests/unit-tests/test_dataloading.py (hunk 19 above)
Comment (zhjwy9343) [reply tests/unit-tests/test_dataloading.py:2670]: added.
